# Page Visits Funnel

In [1]:
import pandas as pd

In [2]:
# Loading the 4 DataFrames
visits = pd.read_csv('visit.csv', parse_dates=[1]) # parse_dates: turn column 1 into real datetime types
cart = pd.read_csv('cart.csv', parse_dates=[1])
checkout = pd.read_csv('checkout.csv', parse_dates=[1])
purchase = pd.read_csv('purchase.csv', parse_dates=[1])

## Task 1: Inspecting the DataFrames

In [3]:
visits.head()

user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00

In [7]:
visits.shape # visits df has 2000 rows and 2 columns

(2000, 2)

In [12]:
cart.head()

user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00

In [11]:
cart.shape # cart df has 348 rows and 2 columns

(348, 2)

In [13]:
checkout.head()

user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 2017-04-20 16:15:00
4  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:52:00

In [14]:
checkout.shape #checkout df has 360 rows and 2 columns

(360, 2)

In [15]:
purchase.head()

user_id       purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85 2017-05-11 04:25:00
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac 2017-09-05 08:45:00
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc 2017-11-20 20:49:00
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d 2017-01-22 15:18:00
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867 2017-06-11 11:32:00

In [16]:
purchase.shape # purchase df has 252 rows and 2 columns

(252, 2)

## Task 2: Combining visits and cart using a left merge.
What percent of users who visited **Cool T-Shirts Inc.** ended up **not placing** a t-shirt in their cart?

In [17]:
# Merge visits and cart
visits_cart = pd.merge(visits, cart, how = 'left')

In [18]:
visits_cart.head()

user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT

In [19]:
visits_cart.shape

(2000, 3)

In [25]:
cart_time_null = len(visits_cart[visits_cart.cart_time.isnull()])
print(f'There are {cart_time_null} timestamps are null in the column cart_time.')

There are 1652 timestamps are null in the column cart_time.


> There are in total *2000* customers visited **CoolTShirts.com**. Out of these *2000* customers, there are *1652* customers that did not add ant t-shirt to their cart.

In [34]:
# Percentage of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart
visits_not_cart_percent = (float(cart_time_null) / len(visits_cart)) * 100
print(f'{visits_not_cart_percent:.2f}% of the users visits CoolTShirts.com but ended up not placing a t-shirt in their cart.')

82.60% of the users visits CoolTShirts.com but ended up not placing a t-shirt in their cart.


## Task 3: Combining cart and checkout
What percentage of users put items in their cart, but did not proceed to checkout?

In [28]:
# Merge cart and checkout
cart_checkout = pd.merge(cart, checkout, how = 'left')

In [29]:
cart_checkout.head()

user_id           cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
1  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
2  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
3  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
4  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   

        checkout_time  
0 2017-11-07 21:14:00  
1 2017-11-07 20:50:00  
2 2017-11-07 21:11:00  
3                 NaT  
4 2017-03-04 11:04:00

In [31]:
cart_checkout.shape

(482, 3)

In [32]:
checkout_time_null = len(cart_checkout[cart_checkout.checkout_time.isnull()])
print(f'There are {checkout_time_null} timestamps are null in the column checkout_time.')

There are 122 timestamps are null in the column checkout_time.


> There are in total *482* customers placed a t-shirt to their cart. Out of these *482* customers, there are *122* customers that did not proceed to checkout.

In [35]:
# Percentage of users who placed a t-shirt in their cart but ended up not checking out
cart_not_checkout_percent = (float(checkout_time_null) / len(cart_checkout)) * 100
print(f'{cart_not_checkout_percent:.2f}% of the users placed a t-shirt in their cart but did not proceed to checkout.')

25.31% of the users placed a t-shirt in their cart but did not proceed to checkout.
